File: leetcode-go/Q001-Q100/Q051_N-Queens/Q051_N-Queens.go
package Q051_N_Queens

func solveNQueens(n int) [][]string {
	board := make([][]byte, n)
	for i := range board {
		board[i] = make([]byte, n)
		for j := range board[i] {
			board[i][j] = '.'
		}
	}
	ans := [][]string{}
	dfs(board, 0, n, &ans)
	return ans
}

func dfs(board [][]byte, row, n int, ans *[][]string) {
	if row == n {
		tmp := make([]string, n)
		for i := 0; i < n; i++ {
			tmp[i] = string(board[i])
		}
		*ans = append(*ans, tmp)
	}
	for i := 0; i < n; i++ {
		if isValid(board, row, i, n) {
			board[row][i] = 'Q'
			dfs(board, row+1, n, ans)
			board[row][i] = '.'
		}
	}
}

func isValid(board [][]byte, x, y, n int) bool {
	for i := 0; i < x; i++ {
		if board[i][y] == 'Q' {
			return false
		}
	}
	for i, j := x-1, y-1; i >= 0 && j >= 0; {
		if board[i][j] == 'Q' {
			return false
		}
		i--
		j--
	}
	for i, j := x-1, y+1; i >= 0 && j < n; {
		if board[i][j] == 'Q' {
			return false
		}
		i--
		j++
	}
	return true
}

File: src/Y2018/backtracking/Q051_N_Queens.java
package Y2018.backtracking;

import java.util.*;

public class Q051_N_Queens {

    Set<Integer> cols = new HashSet<>();
    Set<Integer> diag = new HashSet<>();
    Set<Integer> antiDiag = new HashSet<>();
    List<List<String>> result = new LinkedList<>();

    public List<List<String>> solveNQueens(int n) {
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(board[i], '.');
        }
        dfs(board, 0, n);
        return result;
    }

    private void dfs(char[][] board, int row, int n) {
        for (int col = 0; col < n; col++) {
            if (cols.contains(col)) {
                continue;
            }
            if (diag.contains(row + col)) {
                continue;
            }
            if (antiDiag.contains(row - col)) {
                continue;
            }
            if (row == n - 1) {
                board[row][col] = 'Q';
                convertToList(board);
                board[row][col] = '.';
                return;
            }
            cols.add(col);
            diag.add(row + col);
            antiDiag.add(row - col);
            board[row][col] = 'Q';
            dfs(board, row + 1, n);
            board[row][col] = '.';
            cols.remove(col);
            diag.remove(row + col);
            antiDiag.remove(row - col);
        }
    }

    private void convertToList(char[][] board) {
        List<String> list = new LinkedList<>();
        for (int i = 0; i < board.length; i++) {
            list.add(new String(board[i]));
        }
        result.add(list);
    }

}
